### PR TITLE
[SPARK-18007][SparkR][ML] update SparkR MLP - add initalWeights parameter

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

update SparkR MLP, add initalWeights parameter.
## How was this patch tested?

test added.

### CODE REVIEW

File: R/pkg/R/mllib.R [hunk 1]
```diff
@@ -706,10 +707,13 @@ setMethod("spark.mlp", signature(data = "SparkDataFrame"),
             if (!is.null(seed)) {
               seed <- as.character(as.integer(seed))
             }
+            if (!is.null(initialWeights)) {
+              initialWeights <- as.array(as.numeric(na.omit(initialWeights)))
```
Comment (felixcheung) [inline R/pkg/R/mllib.R]: is the `as.array` required here?

File: R/pkg/R/mllib.R [hunk 2]
```diff
@@ -665,6 +665,7 @@ setMethod("predict", signature(object = "KMeansModel"),
 #' @param tol convergence tolerance of iterations.
 #' @param stepSize stepSize parameter.
 #' @param seed seed parameter for weights initialization.
+#' @param initialWeights initialWeights parameter for weights initialization.
```
Comment (felixcheung) [inline R/pkg/R/mllib.R]: could you clarify that this should be a vector - also is there any restrictions on the length of this?

File: R/pkg/inst/tests/testthat/test_mllib.R [hunk 3]
```diff
@@ -410,6 +410,16 @@ test_that("spark.mlp", {
   model <- spark.mlp(df, layers = c(4, 5, 4, 3), maxIter = 10, seed = 10)
   mlpPredictions <- collect(select(predict(model, mlpTestDF), "prediction"))
   expect_equal(head(mlpPredictions$prediction, 12), c(1, 1, 1, 1, 2, 1, 2, 2, 1, 0, 0, 1))
+
+  # test initialWeights
+  model <- spark.mlp(df, layers = c(4, 3), maxIter = 2, initialWeights =
+    c(0, 0, 0, 0, 0, 5, 5, 5, 5, 5, 9, 9, 9, 9, 9))
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_mllib.R]: if this should be numeric/double, could you add a test with weight as numeric?

File: R/pkg/R/mllib.R [hunk 4]
```diff
@@ -695,7 +696,7 @@ setMethod("predict", signature(object = "KMeansModel"),
 #' @note spark.mlp since 2.1.0
```
Comment (felixcheung) [inline R/pkg/R/mllib.R]: could you update the example with weights

File: R/pkg/R/mllib.R (hunk 1 above)
Comment (WeichenXu123) [inline R/pkg/R/mllib.R]: scala-side wrapper need Array[Double] param so use as.array here, like other parameters such `layers`.

File: R/pkg/R/mllib.R (hunk 2 above)
Comment (WeichenXu123) [inline R/pkg/R/mllib.R]: OK.

File: mllib/src/main/scala/org/apache/spark/ml/r/MultilayerPerceptronClassifierWrapper.scala [hunk 5]
```diff
@@ -73,6 +75,8 @@ private[r] object MultilayerPerceptronClassifierWrapper
       .setStepSize(stepSize)
       .setPredictionCol(PREDICTED_LABEL_COL)
     if (seed != null && seed.length > 0) mlp.setSeed(seed.toInt)
+    if (initialWeights != null) mlp.setInitialWeights(Vectors.dense(initialWeights))
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/MultilayerPerceptronClassifierWrapper.scala]: that seems somewhat involved, so it is likely better to not duplicate the logic on the R side.
we should check here if initialWeights != null its length > 0

**Comment by WeichenXu123** [comment]
@felixcheung
updated following places:
1. clarify `initialWeights` parameter should be a vector
2. add require check in the mlp scala wrapper that if initialWeights != null its length > 0
3. update the example with initial weights

**Comment by felixcheung** [comment]
LGTM.

**Comment by felixcheung** [comment]
merged to master.

**Pull request closed without merging**